hamilton/function_modifiers/macros.py

```diff
@@ -289,3 +303,515 @@ def __init__(self, model_cls, config_param: str, **extra_model_params):
+        """
+        return (
+            (default_namespace,)
+            if self.namespace is ...
+            else (self.namespace,)
+            if self.namespace is not None
+            else ()
+        )
+
+
+def apply(
+    fn, *args: Union[SingleDependency, Any], **kwargs: Union[SingleDependency, Any]
+) -> Applicable:
+    """Applies a function to for a node (or a subcomponent of a node).
+    See documentation for `pipe` to see how this is used.
+
+    :param fn: Function to apply to. Must be validly called as f(**kwargs), and have a 1:1 mapping of kwargs to parameters.
+    :param args: Args to pass to the function -- although these cannot be variable/position-only arguments, they can be
+    positional arguments. If these are not source/value, they will be converted to a value (a literal)
+    :param kwargs: Kwargs to pass to the function. These can be source/value, or they can be literals. If they are literals,
+    they will be converted to a value (a literal)
+    :return: an applicable with the function applied
+    """
+    return Applicable(fn=fn, _resolvers=[], args=args, kwargs=kwargs)
+
+
+class pipe(base.NodeInjector):
+    """Decorator to represent a chained set of transformations. This specifically solves the "node redefinition"
+    problem, and is meant to represent a pipeline of chaining/redefinitions. This is similar (and can happily be
+    used in conjunction with) `pipe` in pandas. In Pyspark this is akin to the common operation of redefining a dataframe
+    with new columns. While it is generally reasonable to contain these constructs within a node's function,
+    you should consider `pipe` for any of the following reasons:
+
+    1.  You want the transformations to display as nodes in the DAG, with the possibility of storing or visualizing
+    the result
+    2. You want to pull in functions from an external repository, and build the DAG a little more procedurally
+    3. You want to use the same function multiple times, but with different parameters -- while `@does`/`@parameterize` can
+    do this, this presents an easier way to do this, especially in a chain.
+
+    To demonstrate the rules for chaining nodes, we'll be using the following example. This is
+    using primitives to demonstrate, but as hamilton is just functions of any python objects, this works perfectly with
+    dataframes, series, etc...
+
+
+    .. code-block:: python
+
+        def _add_one(x: int) -> int:
+            return x + 1
+
+        def _sum(x: int, y: int) -> int:
+            return x + y
+
+        def _multiply(x: int, y: int, z: int=10) -> int:
+            return x * y * z
+
+        @pipe(
+            apply(_add_one),
+            apply(_multiply, y=2),
+            apply(_sum, y=value(3)),
+            apply(_multiply, y=source("upstream_node_to_multiply")),
+        )
+        def final_result(upstream_int: int) -> int:
+            return upstream_int
+
+    .. code-block:: python
```

Comment:
name this code.

Reply:
Didn't know we can, we don't do that anywhere...